### PR TITLE
Public competition page

### LINKED ISSUE #531: Add Public competition list page

### PR DESCRIPTION
**Pull request by Logan-Ruf**
# @ mention of reviewers
@ckcollab @gibsonbailey 

# A brief description of the purpose of the changes contained in this PR.
Create Pagination for Competitions
Create Page to view Public Competitions

# Issues this PR resolves
Resolves #531 

# A checklist for hand testing
- [ ] add checklist here

# Misc. comments
...

# Checklist
- [x] Code review by me 
- [x] Hand tested by me 
- [x] I'm proud of my work
- [x] Code review by reviewer
- [x] Hand tested by reviewer
- [ ] Ready to merge

### CODE REVIEW

File: src/apps/api/views/competitions.py [hunk 1]
```diff
@@ -19,6 +19,8 @@
 from rest_framework.response import Response
 from rest_framework.renderers import JSONRenderer
 from rest_framework_csv.renderers import CSVRenderer
+
+from api.pagination import SmallPagination, LargePagination
```
Comment (Logan-Ruf) [inline src/apps/api/views/competitions.py:23]:
```suggestion
from api.pagination import LargePagination
```
Mobile fix

**Review by Logan-Ruf — COMMENTED** [review]
Mobile review

File: src/apps/api/pagination.py [hunk 2]
```diff
@@ -5,3 +5,15 @@ class BasicPagination(PageNumberPagination):
     page_size = 10
     page_size_query_param = 'page_size'
     max_page_size = 1000
+
+
+class SmallPagination(PageNumberPagination):
+    page_size = 3
+    page_size_query_param = 'page_size'
+    max_page_size = 1000
+
```
[the remark this answers is not in the document]
Comment (ckcollab) [reply src/apps/api/pagination.py:14]: I think remove it because we can just change `page_size` in whatever custom pagination class we're using for testing?

File: src/static/riot/competitions/tile/competition_tile.tag [hunk 3]
```diff
@@ -36,7 +36,7 @@
 
         self.pretty_description = function(description){
             if (description) {
-                description = render_markdown(description)
+                description = render_markdown(description).replaceAll("<p>", "").replaceAll("</p>", "")
```
Comment (ckcollab) [inline src/static/riot/competitions/tile/competition_tile.tag:39]: should we even need this? shouldn't `<p>` markdown get rendered as normal HTML fine? do we need to add `<p>` as a whitelisted tag?

File: src/static/riot/competitions/public-list.tag [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+        <button show="{competitions.previous}" onclick="{handle_ajax_pages.bind(this, -1)}" class="float-left ui inline button active">Back</button>
+        <button hide="{competitions.previous}" disabled="disabled" class="float-left ui inline button disabled">Back</button>
+        { current_page }
+        <button show="{competitions.next}" onclick="{handle_ajax_pages.bind(this, 1)}" class="float-right ui inline button active">Next</button>
+        <button hide="{competitions.next}" disabled="disabled" class="float-right ui inline button disabled">Next</button>
+    </div>
+<script>
+    var self = this
+    self.competitions = {}
+    self.competitions_cache = {}
```
Comment (ckcollab) [inline src/static/riot/competitions/public-list.tag:45]: nice :)

File: src/static/riot/competitions/public-list.tag [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+                    <div class="ui divider"></div>
+                    <strong>{competition.participant_count}</strong> Participants
+                </div>
+            </div>
+        </a>
+    </div>
+    <div class="pagination-nav" hide="{get_array_length(competitions.results) === competitions.count}">
+        <button show="{competitions.previous}" onclick="{handle_ajax_pages.bind(this, -1)}" class="float-left ui inline button active">Back</button>
+        <button hide="{competitions.previous}" disabled="disabled" class="float-left ui inline button disabled">Back</button>
+        { current_page }
```
Comment (ckcollab) [inline src/static/riot/competitions/public-list.tag:38]: should we say "page 1 of X" or something here? I saw a "1" floating and it was kinda confusing at first

File: src/static/riot/competitions/tile/competition_tile.tag (hunk 3 above)
Comment (Logan-Ruf) [reply src/static/riot/competitions/tile/competition_tile.tag:39]: It doesn't seem like it does. This is the current [codabench](www.codabench.org) page
image

Comment (ckcollab) [reply src/static/riot/competitions/tile/competition_tile.tag:39]: is the "rendering of markdown" even working then? does that description get rendered properly elsewhere?

Comment (Logan-Ruf) [reply src/static/riot/competitions/tile/competition_tile.tag:39]: It looks like the `render_markdown` was actually unnecessary in the first place. Updated the code.

File: src/apps/api/pagination.py [hunk 6]
```diff
@@ -11,3 +12,12 @@ class LargePagination(PageNumberPagination):
     page_size = 20
     page_size_query_param = 'page_size'
     max_page_size = 1000
+
+    def get_paginated_response(self, data):
```
Comment (ckcollab) [inline src/apps/api/pagination.py:16]: why is this needed?

File: src/static/riot/competitions/public-list.tag [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+            self.competitions = self.competitions_cache[self.current_page]
+            history.pushState("", document.title, "?page="+self.current_page)
+            self.update()
+        } else {
+            return CODALAB.api.get_public_competitions({"page":self.current_page})
+                .fail(function (response) {
+                    toastr.error("Could not load competition list")
+                })
+                .done(function (response){
+                    console.log(response)
```
Comment (ckcollab) [inline src/static/riot/competitions/public-list.tag:67]: pluck
